src/range.rs

```diff
@@ -0,0 +1,355 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[{}..{})", self.start(), self.end())
+    }
+}
+
+/// Identity methods.
+impl TextRange {
+    /// The start point of this range.
+    pub const fn start(self) -> TextSize {
+        self.start
+    }
+
+    /// The end point of this range.
+    pub const fn end(self) -> TextSize {
+        self.end
+    }
+
+    /// The size of this range.
+    ///
+    /// # Panics
+    ///
+    /// When `end() < start()`, triggers a subtraction overflow.
+    /// This will panic with debug assertions, and overflow without.
+    pub const fn len(self) -> TextSize {
+        // HACK for const fn: math on primitives only
+        TextSize(self.end().raw - self.start().raw)
+    }
+
+    /// Check if this range empty or reversed.
+    ///
+    /// When `end() < start()`, this returns false.
+    /// Code should prefer `is_empty()` to `len() == 0`,
+    /// as this safeguards against incorrect reverse ranges.
+    pub const fn is_empty(self) -> bool {
+        // HACK for const fn: math on primitives only
+        self.start().raw >= self.end().raw
+    }
+}
+
+/// Manipulation methods.
+impl TextRange {
+    /// A range covering the text size of some text-like object.
+    pub fn of(size: impl TextSized) -> TextRange {
+        TextRange(0.into(), size.text_size())
+    }
+
+    /// An empty range at some text size offset.
+    pub fn at(size: impl Into<TextSize>) -> TextRange {
+        let size = size.into();
+        TextRange(size, size)
+    }
+
+    /// Set the length without changing the starting offset.
+    pub fn with_len(self, len: impl Into<TextSize>) -> TextRange {
+        TextRange(self.start(), self.start() + len.into())
+    }
+
+    /// Set the starting offset without changing the length.
+    pub fn with_offset(self, offset: impl Into<TextSize>) -> TextRange {
+        TextRange::at(offset).with_len(self.len())
+    }
+
+    /// Check if this range completely contains another range.
+    pub fn contains(self, other: TextRange) -> bool {
+        self.start() <= other.start() && other.end() <= self.end()
+    }
+
+    /// The range covered by both ranges, if it exists.
+    /// If the ranges touch but do not overlap, the output range is empty.
+    pub fn intersection(lhs: TextRange, rhs: TextRange) -> Option<TextRange> {
+        let start = cmp::max(lhs.start(), rhs.start());
+        let end = cmp::min(lhs.end(), rhs.end());
+        Some(TextRange(start, end)).filter(|_| start <= end)
+    }
+
+    /// The smallest range that completely contains both ranges.
+    pub fn covering(lhs: TextRange, rhs: TextRange) -> TextRange {
+        let start = cmp::min(lhs.start(), rhs.start());
+        let end = cmp::max(lhs.end(), rhs.end());
+        TextRange(start, end)
+    }
+
+    /// Check if this range contains a point.
+    ///
+    /// The end index is considered excluded.
+    pub fn contains_point(self, point: impl Into<TextSize>) -> bool {
```

Comment:
I feel pretty strong, here and elsewhere, that we should not use `Into`s. 

The general argument is that, while `Into` do make API more fluent to use with *literal* values, they lead to more confusing type errors (and sometimes to new type errors) when type inference is involved. Besides, not-generic APIs are simpler.

The specific point here is that raison d'être of this library is to separate, on the type level, string offsets from all other numbers, and allowing `range + 92` goes exactly against that. 

Additionally, this library should really never be used with literal numbers, outside of tests. Instead of `range + 1`, one should write `range + TextSize::of(' ')` or whatever specific character one wants to shift the range by.